src/textual/_xterm_parser.py

```diff
@@ -129,25 +131,42 @@ def parse(self, on_token: TokenCallback) -> Generator[Awaitable, str, None]:
                         if len(peek_buffer) == 1 and not more_data():
                             continue
 
+                # Look ahead through the suspected escape sequence for a match
                 while True:
-                    sequence += yield read1()
+
+                    # If we run into another ESC at this point, then we've failed
+                    # to find a match, and should issue everything we've seen within
+                    # the suspected sequence as Key events instead.
+                    buffer = yield self.peek_buffer()
+                    if (
+                        buffer
+                        and buffer[0] == ESC
+                        or len(sequence) > _MAX_SEQUENCE_SEARCH_THRESHOLD
+                    ):
+                        for character in sequence:
```

Comment:
There's some very similar code further down, I wonder if it could be hoisted in to a closure?

Reply:
I considered this, but will do it now that I know I'm not the only one with that thought :)

Reply:
Oh, you said closure...

I've factored it out into a separate method